Add normative wording and use set-cookie value format (#20)
w3c/correlation-context: correlation_context/HTTP_HEADER_FORMAT_RATIONALE.md

# Correlation Context Header Format Rationale

This document provides a rationale for the decisions made for the `Correlation-Context` header format.

## General considerations

- It should be human-readable. Cryptic headers would hide the fact of potential information disclosure.
- It should be append-able (comma-separated) https://tools.ietf.org/html/rfc7230#page-24 so nodes
can add context properties without parsing existing headers.
- Keys are a single word in ASCII, and values should be a short string in UTF-8 or a derivative of an URL.

## Why a single header?

Another option would be to use prefixed headers, such as `trace-context-X`, where `X` is a propagated
field name. That could reduce the size of the data, particularly in http/2, where header
compression can apply.

Generally speaking, a `Correlation-Context` header may be split into multiple headers, and
compression may be at the same ballpark as repeating values are converted into a single value
in HPAC's dynamic collection. That said, there was no profiling made to make this decision.

The approach with multiple headers has the following problems:

> - Name values limitation is much more pressing when the context name is used a part of a header

Reviewer: Not sure what "Name values" means. We should rephrase.

Author: https://github.com/w3c/correlation-context/issues/26

name.

> - The comma-separated format similar to the proposed still needs to be supported in every individual header which makes parsing harder.

Reviewer: I don't follow this point, I don't think it's technically correct - comma-separated format is not required for headers if they are meant to be semantically unique.

Author: https://github.com/w3c/correlation-context/issues/26

> - A single header is more comfortable to configure for tracing by many app servers.

Reviewer: add CORS and header whitelisting as a reason?

Reviewer 2: ```suggestion
- A single header is more comfortable to configure for tracing by many app servers, and makes CORS whitelisting more straightforward.
```

## Why not Vary-style?

> The [Vary](https://tools.ietf.org/html/rfc7231#section-7.1.4) approach is another alternative,

Reviewer: ```suggestion
The [Vary](https://tools.ietf.org/html/rfc7231#section-7.1.4) approach is another alternative,
where a fixed header is used to enumerate the list of other headers that actually contain the data.
```

> which could be used to accomplish the same. For example, `Correlation-Context: x-b3-parentid;

Reviewer: `Correlation-Context: x-b3-parentid;` feels like a bad/confusing example, since baggage is not meant to be used to pass tracing system's fields. We we change to something more business specific? E.g.
```suggestion
which could be used to accomplish the same. For example, `Correlation-Context: x-department; x-user-id;
```

> ttl=1` could tell the propagation to look at and forward the parent ID header, but only to the

Reviewer: this part makes even less sense - the interpretation of ttl property is completely unrelated to the discussion of Vary header format.

Author: https://github.com/w3c/correlation-context/issues/26

next hop. This has an advantage of HTTP header compression (hpack) and also weave-in with legacy
tracing headers.

Vary approach may be implemented as a new "header reference" value type `ref`.
`Correlation-Context: x-b3-parentid;type=ref;ttl=1` if proven needed.

> 

Reviewer: ironically, this section does not provide any explanation why Vary was not chosen (the reasons are similar to CTX-X format in the previous section)

## Trimming of spaces

The header should be human-readable and editable. Thus spaces are allowed before and after the comma, equal sign, and semicolon separators. It makes manual editing of headers less error-prone. It also allows better visual separation of fields when value modified manually.

> ## Case sensitivity of names

Reviewer: ```suggestion
## Case sensitivity of keys
```

Reviewer 3: I would consider just removing the justification for case sensitive keys entirely

There are few considerations why the names should be case sensitive:

> - some keys may be a URL query string parameters which are case sensitive

Reviewer: don't know what this means, doesn't sound like a good argument.

Why do we need to justify case-sensitive keys?

Author: This seems to be historic and part of the rationale for completeness. Happy to remove it, if @SergeyKanzhelev agrees as well as he most likely wrote it.

Author: https://github.com/w3c/correlation-context/issues/26

- Name values limitation is much more pressing when the context name is used a part of a header
name.
- The comma-separated format similar to the proposed still needs to be supported in every individual header which makes parsing harder.

> - A single header is more comfortable to configure for tracing by many app servers, and makes CORS whitelisting more straightforward.

Reviewer 3: I would even use stronger wording and say it is a prerequisite if we want to be CORS whitelisted.

Author: This is not about getting the header approved. It is referring to whitelist this on the server and of course, you can do that with multiple headers. It's just more work.
